src/scripts/components/header/SearchAutocomplete.js

```diff
@@ -0,0 +1,127 @@
+import React from 'react';
+import classNames from 'classnames';
+import SurahsStore from 'stores/SurahsStore';
+import connectToStores from 'fluxible-addons-react/connectToStores';
+import request from 'superagent';
+import Settings from 'constants/Settings';
+
+class SearchAutocomplete extends React.Component {
+  constructor(props, context) {
+    super(props, context);
+    this.state = { value: '', ayat: [], surahs: [] };
+    this.cached = {};
+  }
+  componentWillReceiveProps( nextProps ) {
+    //console.log('searchautocomplete componentWillReceiveProps', nextProps);
+    this.setState({ value: nextProps.value });
+    this.suggest( nextProps.value );
+  };
+
+  suggest(value) {
+    if (value.length >= 2) {
+      this.handleSurahSuggestions(value);
+      if (value.length >= 3) {
+        this.handleAyahSuggestions(value);
+      }
+      else {
+        this.setState({ ayat: [] });
+      }
+    }
+    else {
+      this.setState( { surahs: [] } );
+    }
+  };
+
+  handleSurahSuggestions(value) {
+    var escaped = value.replace(/[-\\^$*+?.()|[\]{}]/g, "\\$&");
+    var matches = [];
+    this.props.surahs.forEach( function( surah ) {
+      if ( RegExp( escaped, "i" ).test( surah.name.simple.replace( /['-]/g, '' )  ) ) {
+        matches.push( [ surah.name.simple, surah.id ] );
+      } else
+      if ( RegExp( escaped, "i" ).test( surah.name.arabic ) ) {
+        matches.push( [ surah.name.arabic, surah.id ] );
+      }
+    } );
+    matches = matches.sort(function( a, b ) {
+      return a[1] < b[1] ? -1 : 1; // <-- order by surah id
+      return RegExp( '^'+ escaped, 'i' ).test( a[0] ) ? -1 : 1; // <-- order by whether it starts the same
+    }).map( function( s ) {
+      // var text = s[0].replace(RegExp(escaped, "gi"), "<mark>$&</mark>"); // <-- this will highlight the matched part
+      var text = s[0];
+      return { text: '<b>'+ text +'</b>', href: '/'+ s[1] };
+    } ).slice(0,5);
+    this.setState( { surahs: matches } );
+  };
+
+  handleAyahSuggestions(value) {
+    var self = this;
```

Comment:
you won't need `self` if you use arrow functions at the bottom. So `.end((err, res) => {})`